organic_scoring/organic_scoring_base.py

```diff
@@ -0,0 +1,316 @@
+        self._trigger_scaling_factor = trigger_scaling_factor
+        assert self._trigger_scaling_factor > 0, "The scaling factor must be higher than 0."
+        self._organic_queue = organic_queue
+        if self._organic_queue is None:
+            self._organic_queue = OrganicQueue()
+        self._step_counter = 0
+        self._step_lock = asyncio.Lock()
+
+        # Bittensor's internal checks require synapse to be a subclass of bt.Synapse.
+        # If the methods are not overridden in the derived class, None is passed.
+        self._axon.attach(
+            forward_fn=self._on_organic_entry,
+            blacklist_fn=self._blacklist_fn if is_overridden(self._blacklist_fn) else None,
+            priority_fn=self._priority_fn if is_overridden(self._priority_fn) else None,
+            verify_fn=self._verify_fn if is_overridden(self._verify_fn) else None,
+        )
+
+
+    def increment_step(self):
+        """Increment the step counter if the trigger is set to `steps`."""
+        with self._step_lock:
+            if self._trigger == "steps":
+                self._step_counter += 1
+
+    def set_step(self, step: int):
+        """Set the step counter to a specific value.
+
+        Args:
+            step: The step value to set.
+        """
+        with self._step_lock:
+            if self._trigger == "steps":
+                self._step_counter = step
+
+    @abstractmethod
+    async def _on_organic_entry(self, synapse: bt.Synapse) -> bt.Synapse:
+        """Handle an organic entry.
+
+        Important: this method must add the required values to the `_organic_queue`.
+
+        Args:
+            synapse: The synapse to handle.
+
+        Returns:
+            bt.StreamingSynapse: The handled synapse.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    async def _query_miners(self, sample: Any) -> dict[str, Any]:
+        """Query the miners with a sample
+
+        Args:
+            sample: The sample to query with.
+
+        Returns:
+            dict[str, Any]: The responses from the miners.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    async def _generate_rewards(self, sample: Any, responses: Sequence[Any], reference: Any = None) -> dict[str, Any]:
+        """Generate rewards based on the sample and responses.
+
+        Args:
+            sample: The sample to use.
+            responses: The responses from the miners.
+            reference: The reference data. Defaults to None.
+
+        Returns:
+            dict[str, Any]: The generated rewards information.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    async def _set_weights(self, rewards: dict[str, Any]):
+        """Set the weights for the miners.
+
+        Args:
+            rewards: Dict with rewards and any additional info.
+        """
+        raise NotImplementedError
+
+    async def _generate_reference(self, sample: Any) -> Optional[Any]:
```

Comment:
is it correct for this not to be an abstract method?

Reply:
Because it's optional